### ISSUE TITLE
annotate_special/source.set_data: data tensor indexing edge case

### ISSUE THREAD

**Issue by jordan-palmer** [FlamTeam/flamedisx#224]
**Context**
I came across an issue using the `annotate_special` function.

When you call `source.set_data` and your cuts remove some number of events, we wish the data tensor with real events such that the total is a multiple of the batch size. In FD, we seem to use the 0th index, which is fine. This is done here
```                
self.data = pd.concat([self.data, df_pad], ignore_index=True)
```
where `df_pad` are the 0th event repeated some number of times. This line is only called if `n_padding` is > 0. In the case where `n_padding = 0` AND the case where the number of events cut are multiples of the batch size (10,20,30,.. in the default case) then this "ignore index" is not called and the final data tensor is awkwardly indexed. In the default annotate methods, this is does not cause an error, but this scenario does of course happen. Would be interesting to know if this could cause a feature somewhere? 

**Error message**

The result of this feature can break the `annotate_special` method with the following error:

```
/content/FlameNEST/flamedisx/nest/lxe_blocks/quanta_splitting.py in <listcomp>(.0)
    353         max_num_energies = max(min(len(self.source.energies), self.source.max_dim_sizes['energy']), 2)
    354         # Pad with 0s at the end to make each one the same size
--> 355         [bounds.extend([0]*(max_num_energies - len(bounds))) for bounds in d['ions_produced_min'].values]
    356 
    357         return True
AttributeError: 'float' object has no attribute 'extend'
```
After some researching, this error is due to this indexing issue. This occurs even with a fixed MC reservoir.

**Proposed solutions** 

- Update annotate_special to not break with awkward indexing
- if the `n_padding` = 0, just call `self.data = self.data.reset_index(drop=True)`. I don't believe this causes any issues in the logic from a couple of tests I have done. But this is by far the quickest solution and I always think it is good practise to reset the indices of dfs. 

**Recreate this error**

```
Likelihood = fd.LogLikelihood(sources = dict(er=fd.lux.LUXERSource),
                              free_rates = ('er'),
                              arguments = dict(er={'energy_min':0.01,'energy_max':20.}),
                             )
i=0
while i < 20:
    print('set data number ',i)
    df_data = fd.lux.LUXERSource(energy_min = 0.01, energy_max= 20.).simulate(200)
    Likelihood.set_data(df_data)
    i+=1
```

Shouldn't take more than 10 to break with this error. To clearly see that this error is occurring where I say it is, you can print `self.data.index` before and after the if loop  `if not _skip_tf_init:` and compare them. If the `self.data.index` AFTER the if statement does not have the form `RangeIndex(start=0, stop=200, step=1)` but something like 
```
Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
            ...
            190, 191, 192, 193, 194, 195, 196, 197, 198, 199],
           dtype='int64', length=190)
```
it should break.

**Comment by JelleAalbers** [comment]
Yes, the code here https://github.com/FlamTeam/flamedisx/blob/f626fe5c7dde0abbc2426333bbee82a898abd059 relies on the dataframe's index. I usually avoid this side of pandas and e.g. stick with `.iloc` instead of `.loc`.

Maybe you could collect the min/max bounds in a list here, then assign it to a dataframe column after the loop? But I haven't fully grokked the code here, so I could well be missing some complication.

I like the idea of resetting the index in `set_data`; I think you could do that whether or not the padding is 0.

**Comment by plt109** [comment]
I failed to recreate the error with `fd.SR1ERSource` with the current main (`7da5d4622a5a58d63c20d15eaa4dc3c23c29dcbc`).
<img width="571" alt="Screen Shot 2022-05-16 at 12 07 10 AM" src="https://user-images.githubusercontent.com/46324102/168495917-4e8ceb27-745e-4b6a-86c3-94524ad00fa1.png">

I couldn't access the page that @JelleAalbers mentioned (https://github.com/FlamTeam/flamedisx/blob/f626fe5c7dde0abbc2426333bbee82a898abd059), but I vaguely remember some discussion about `.iloc` vs `.loc` in the NEST pre-quanta PR.

Random question out of sheer curiosity:
Under what circumstances would one have to set data to the same likelihood multiple times?

**Comment by jordan-palmer** [comment]
> Yes, the code here https://github.com/FlamTeam/flamedisx/blob/f626fe5c7dde0abbc2426333bbee82a898abd059 relies on the dataframe's index. I usually avoid this side of pandas and e.g. stick with `.iloc` instead of `.loc`.
> 
> 
> 
> Maybe you could collect the min/max bounds in a list here, then assign it to a dataframe column after the loop? But I haven't fully grokked the code here, so I could well be missing some complication.
> 
> 
> 
> I like the idea of resetting the index in `set_data`; I think you could do that whether or not the padding is 0.

Yeah this is right
Resetting the index will prevent issues in the other sources once the Bayes bounds have been added to them. But I agree, we should avoid using `.loc` for this reason.

**Comment by jordan-palmer** [comment]
> I failed to recreate the error with `fd.SR1ERSource` with the current main (`7da5d4622a5a58d63c20d15eaa4dc3c23c29dcbc`).
> 
> <img width="571" alt="Screen Shot 2022-05-16 at 12 07 10 AM" src="https://user-images.githubusercontent.com/46324102/168495917-4e8ceb27-745e-4b6a-86c3-94524ad00fa1.png">
> 
> 
> 
> I couldn't access the page that @JelleAalbers mentioned (https://github.com/FlamTeam/flamedisx/blob/f626fe5c7dde0abbc2426333bbee82a898abd059), but I vaguely remember some discussion about `.iloc` vs `.loc` in the NEST pre-quanta PR.
> 
> 
> 
> Random question out of sheer curiosity:
> 
> Under what circumstances would one have to set data to the same likelihood multiple times?

You can only recreate this issue with the NEST sources since the others don't use the Bayes bounds functions yet. But yes, this is exactly where the issue comes from. I just like a data frame which is correctly indexed and all sources can sometimes produce incorrectly indexed data frames, it just rears it's head in the NEST sources (for now).

Random qn answer:
You would fit to the same likelihood multiple times for pull studies since you want to change the dataset each time but the sources remain the same so no need to update the likelihood

**Comment by jordan-palmer** [comment]
PR is here https://github.com/FlamTeam/flamedisx/pull/225

**Comment by robertsjames** [comment]
Thanks for the detailed summary Jordan! I'll take a look at your PR as soon as I can

**Comment by JelleAalbers** [comment]
Fixed by Jordan's PR #225, thanks!
